extensions/azurecore/src/account-provider/auths/azureDeviceCode.ts

```diff
@@ -0,0 +1,159 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as azdata from 'azdata';
+import * as vscode from 'vscode';
+import * as nls from 'vscode-nls';
+import {
+	AzureAuth,
+	TokenClaims,
+	AccessToken,
+	RefreshToken
+
+} from './azureAuth';
+
+import {
+	AzureAccountProviderMetadata,
+	AzureAccount,
+	AzureAuthType,
+	// Tenant,
+	// Subscription
+} from '../interfaces';
+
+import { SimpleTokenCache } from '../simpleTokenCache';
+const localize = nls.loadMessageBundle();
+
+interface DeviceCodeLogin { // https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-device-code
+	device_code: string,
+	expires_in: number;
+	interval: number;
+	message: string;
+	user_code: string;
+	verification_url: string
+}
+
+interface DeviceCodeLoginResult {
+	token_type: string,
+	scope: string,
+	expires_in: number,
+	access_token: string,
+	refresh_token: string,
+}
+
+export class AzureDeviceCode extends AzureAuth {
```

Comment:
Have you gone through a security review for this? I'm worried that running our own web server like this is going to be considered a security risk for someone to hijack a users auth tokens.

Reply:
I haven't but this is currently the feature that has already shipped. (We're running a webserver right now).

It's also how vscode handles oauth. So I don't think it'll be an issue.

I think if we want a security review it should technically happen independent of this PR.

Reply:
That's good that VS Code is already doing this - but yeah we definitely need to do a security review at some point. Let me know if you want help getting that set up.

Reply:
Sure thing!